diff/diff.go

```diff
@@ -0,0 +1,276 @@
+	}
+}
+
+// Differ is a type that can diff values.  It keeps its own
+// diffing style.
+type Differ struct {
+	wrappers []func(string) string
+
+	actual   *Differ
+	expected *Differ
+}
+
+// New creates a Differ, using the passed in opts to manipulate
+// its diffing behavior and output.
+//
+// If opts is empty, the default options used will be:
+// [ WithFormat(">%s<"), Actual(WithFormat("%s!=")) ]
+func New(opts ...Opt) *Differ {
+	d := Differ{}
+	if len(opts) == 0 {
+		opts = []Opt{WithFormat(">%s<"), Actual(WithFormat("%s!="))}
+	}
+	for _, opt := range opts {
+		d = opt(d)
+	}
+	return &d
+}
+
+// format is used to format a string using the wrapper functions.
+func (d Differ) format(v string) string {
+	for _, w := range d.wrappers {
+		v = w(v)
+	}
+	return v
+}
+
+// Diff takes two values and returns a string showing a
+// diff of them.
+func (d *Differ) Diff(actual, expected interface{}) string {
+	return d.diff(reflect.ValueOf(actual), reflect.ValueOf(expected))
+}
+
+func (d *Differ) genDiff(format string, actual, expected interface{}) string {
+	afmt := fmt.Sprintf(format, actual)
+	if d.actual != nil {
+		afmt = d.actual.format(afmt)
+	}
+	efmt := fmt.Sprintf(format, expected)
+	if d.expected != nil {
+		efmt = d.expected.format(efmt)
+	}
+	return d.format(afmt + efmt)
+}
+
+func (d *Differ) diff(av, ev reflect.Value) string {
+	if !av.IsValid() {
+		if !ev.IsValid() {
+			return "<nil>"
+		}
+		return d.genDiff("%v", "<nil>", ev.Interface())
+	}
+	if !ev.IsValid() {
+		return d.genDiff("%v", av.Interface(), "<nil>")
+	}
+
+	if av.Kind() != ev.Kind() {
+		return d.genDiff("%T", av.Interface(), ev.Interface())
+	}
+
+	switch av.Interface().(type) {
+	case []rune, []byte, string:
+		// we want to find differences in the middle of strings and
+		// string-like types, whenever possible.
+		if av.Len() != ev.Len() {
+			break // let the default logic handle this
+		}
+
+		strTyp := reflect.TypeOf("")
+		var curra, currb, out string
+		for i := 0; i < av.Len(); i++ {
+			match := av.Index(i).Interface() == ev.Index(i).Interface()
+			if !match {
+				curra += av.Index(i).Convert(strTyp).Interface().(string)
+				currb += ev.Index(i).Convert(strTyp).Interface().(string)
+				continue
+			}
+			if len(curra) > 0 {
+				out += d.genDiff("%s", curra, currb)
+				curra, currb = "", ""
+			}
+			out += av.Index(i).Convert(strTyp).Interface().(string)
+		}
+		if len(curra) > 0 {
+			out += d.genDiff("%s", curra, currb)
+			curra, currb = "", ""
+		}
+		return out
+	}
+
+	switch av.Kind() {
+	case reflect.Ptr, reflect.Interface:
+		return d.diff(av.Elem(), ev.Elem())
+	case reflect.Slice, reflect.Array:
+		if av.Len() != ev.Len() {
+			// TODO: do a more thorough diff of values
+			return d.genDiff(fmt.Sprintf("%T(len %%d)", av.Interface()), av.Len(), ev.Len())
+		}
+		var elems []string
+		for i := 0; i < av.Len(); i++ {
+			elems = append(elems, d.diff(av.Index(i), ev.Index(i)))
+		}
+		return "[ " + strings.Join(elems, ", ") + " ]"
```

Comment:
I should probably go and check on how `%v` prints slices.  I wasn't specifically going for json-like, but you're right that it looks a bit like that.  I was just intending it to look human-readable (which json is usually not).

I think I'd _rather_ avoid json, simply because I expect that we'll want to refine the output as we use it more, and explicitly using json implies intent that we (or, at least, I) don't really have.